src/PortingAssistant.Client.Telemetry/Uploader.cs

```diff
@@ -12,130 +10,308 @@
 using System.Net.Http;
 using System.Net;
 using System.Text;
+using System.Threading;
 using System.Threading.Tasks;
-using Amazon;
 
 namespace PortingAssistant.Client.Telemetry
 {
     public class Uploader
     {
-        public static bool Upload(TelemetryConfiguration teleConfig, string profile, ITelemetryClient teleClient)
+        private readonly TelemetryConfiguration _configuration;
+        private readonly ITelemetryClient _client;
+        private Dictionary<string, int> _fileLineNumberMap = new();
+        private readonly Dictionary<string, int> _updatedFileLineNumberMap = new();
+        private readonly ILogger _logger;
+        private string _lastReadTokenFile;
+        private readonly Dictionary<string, int> _errors = new();
+        private const long LogDirectorySizeLimit = 250000000;
+        private readonly bool _shareMetrics;
+
+        public Uploader(TelemetryConfiguration telemetryConfig,
+            ITelemetryClient telemetryClient,
+            ILogger logger,
+            bool shareMetrics)
+        {
+            _configuration = telemetryConfig;
+            _client = telemetryClient;
+            _logger = logger;
+            _shareMetrics = shareMetrics;
+            ReadFileLineMap();
+            GetLogName = GetLogNameDefault;
+        }
+
+        /// <summary>
+        /// Takes file name and returns log name for upload. Returning blank will skip the file
+        /// </summary>
+        public Func<string, string> GetLogName { get; set; }
+
+        public bool Run()
         {
             try
             {
-                var fileEntries = Directory.GetFiles(teleConfig.LogsPath)
-                    .Where(f =>
-                        Path.GetFileName(f)
-                            .StartsWith("portingAssistant-client-cli") &&
-                        teleConfig.Suffix.ToArray().Any(x => f.EndsWith(x))
-                    ).ToList();
-
-                var lastReadTokenFile = Path.Combine(teleConfig.LogsPath, "lastToken.json");
-                var fileLineNumberMap = new Dictionary<string, int>();
-                if (File.Exists(lastReadTokenFile))
+                if (_shareMetrics)
                 {
-                    fileLineNumberMap = JsonConvert.DeserializeObject<Dictionary<string, int>>(File.ReadAllText(lastReadTokenFile));
+                    Upload();
                 }
+                CleanupLogFolder();
```

Comment:
This will be called everytime we upload logs, maybe we can call Cleanup just once when the app starts.